### ISSUE TITLE
[checkpoint] Switch away from pickle-base serialization

### ISSUE THREAD

**Issue by kumpera** [pytorch/pytorch#76441]
### 🚀 The feature, motivation and pitch

We currently use pickle for serializing everything.
Beyond efficiency issues, it makes backwards compatibility quite hard.

This proposal is to switch entirely to a flatbuffers based serialization.
There's a precedent in the mobile front and it's a reasonable format to begin with.

**Comment by vadimkantorov** [comment]
Related? https://github.com/pytorch/pytorch/issues/52181

**Comment by kumpera** [comment]
> Related? #52181

Yes, somewhat. There are a few issue we face with distributed checkpointing and using torch.load / torch.save beyond those inherent to pickle. In no particular order:

Split tensor metadata from data and save them separately.
Have an efficient implementation of partial loading a tensor. IE `torch.load(...).view(...)`

**Comment by vadimkantorov** [comment]
> Have an efficient implementation of partial loading a tensor.

I wonder if HDF5 allows that...

**Comment by vadimkantorov** [comment]
Maybe also related: https://github.com/pytorch/pytorch/issues/76924
